### PR TITLE
Added Develocity plugin for build scans and remote build cache

### PR DESCRIPTION
**Pull request by Mr3zee**
**Subsystem**
Gradle

**Motivation**
Publish build scans and build cache

**Solution**
Added Develocity configs

### CODE REVIEW

File: gradle-settings-conventions/src/main/kotlin/customization.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            }
+
+            if (gradle.rootProject.hasProperty(buildId)) {
+                ge.buildScan.value("CI build id", gradle.rootProject.property(buildId) as String)
+            }
+        }
+
+        if (!isCIRun) {
+            // Git commit id
+            val commitId = execute("git rev-parse --verify HEAD")
```
Comment (osipxd) [inline gradle-settings-conventions/src/main/kotlin/customization.kt:34]: **Nit:** Maybe we should move it to a separate function with a different name? Like `enrichGitData`. Or rename the current function.

File: docs/develocity.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Develocity usage guide
+
+> Build scans and remote build cache are for JetBrains developers only
```
Comment (osipxd) [inline docs/develocity.md:3]:
```suggestion
> [!NOTE]
> Build scans and remote build cache are for JetBrains developers only.
```
> [!NOTE]
> Build scans and remote build cache are for JetBrains developers only.

File: settings.gradle.kts [hunk 3]
```diff
@@ -3,31 +3,21 @@
...
-rootProject.name = "ktor"
-
-val CACHE_USER = System.getenv("GRADLE_CACHE_USER")
-
-if (CACHE_USER != null) {
-    val CACHE_PASSWORD = System.getenv("GRADLE_CACHE_PASSWORD")
-    buildCache {
-        remote(HttpBuildCache::class) {
-            isPush = true
-            setUrl("https://ktor-gradle-cache.teamcity.com/cache/")
```
Comment (osipxd) [inline settings.gradle.kts:22]: Can't find this link in the new configuration. Is it hidden on the Develocy server?

**Review by osipxd — APPROVED** [review]
Wow! 🚀

File: gradle-settings-conventions/src/main/kotlin/customization.kt (hunk 1 above)
Comment (Mr3zee) [reply gradle-settings-conventions/src/main/kotlin/customization.kt:34]: Sure, that'd be better

**Review by bjhham — APPROVED** [review]

File: gradle-settings-conventions/src/main/kotlin/conventions-develocity.settings.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+        capture {
+            fileFingerprints = true
+        }
+
+        buildScanPublished {
+            scanJournal.appendText("${Date()} — $buildScanUri — $startParameter\n")
+        }
+
+        publishing.onlyIf { true }
```
Comment (Goooler) [inline gradle-settings-conventions/src/main/kotlin/conventions-develocity.settings.gradle.kts:34]:
```suggestion
        publishing.onlyIf { isCIRun }
```

File: gradle-settings-conventions/src/main/kotlin/params.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+const val DEVELOCITY_SERVER = "https://ge.jetbrains.com"
+const val GITHUB_REPO = "https://github.com/ktorio/ktor"
+const val TEAMCITY_URL = "https://ktor.teamcity.com"
+
+val isCIRun = System.getenv("TEAMCITY_VERSION") != null
```
Comment (Goooler) [inline gradle-settings-conventions/src/main/kotlin/params.kt:9]:
```suggestion
val isCIRun = System.getenv("TEAMCITY_VERSION") != null || System.getenv("CI") != null
```

Comment (Goooler) [reply gradle-settings-conventions/src/main/kotlin/params.kt:9]: `CI` env variable is available for GitHub Actions.

File: gradle-settings-conventions/src/main/kotlin/conventions-develocity.settings.gradle.kts (hunk 4 above)
Comment (Mr3zee) [reply gradle-settings-conventions/src/main/kotlin/conventions-develocity.settings.gradle.kts:34]: We want to publish local CI builds. The change I would is to introduce a property that can disable it for external collaborators

File: gradle-settings-conventions/src/main/kotlin/params.kt (hunk 5 above)
Comment (Mr3zee) [reply gradle-settings-conventions/src/main/kotlin/params.kt:9]: We don't run builds on GH Actions

**Review by e5l — APPROVED** [review]

**Comment by Mr3zee** [comment]
Build cache publishing works, so I'm merging it

**Pull request merged**